File: frontend/src/features/examinations/lib/examination-lock.ts
/**
 * BUG-033 / S02 examination edit locks (server-persisted status only).
 *
 * - confirmed: full form lock
 * - completed without revision history: first-pass completion seal (results + delete)
 * - completed with revision history: post-unconfirm working copy (editable)
 *
 * Accepts JA labels (form/transform) and EN API enums so a missed transform
 * cannot silently unlock a sealed exam.
 */

export type ExaminationLockStatus =
  | "依頼中"
  | "検査中"
  | "結果入力済み"
  | "完了"
  | "確定";

/** Normalize FE JA labels and BE EN enums used by lock checks. */
export function normalizeExaminationLockStatus(
  status: string | undefined | null,
): ExaminationLockStatus | string | undefined | null {
  if (status == null || status === "") return status;
  switch (status) {
    case "pending":
      return "依頼中";
    case "in_progress":
      return "検査中";
    case "result_entered":
      return "結果入力済み";
    case "completed":
    case "完了":
      return "完了";
    case "confirmed":
    case "確定":
      return "確定";
    default:
      return status;
  }
}

export function isPersistedConfirmedStatus(
  status: string | undefined | null,
): boolean {
  return normalizeExaminationLockStatus(status) === "確定";
}

/** First-pass「完了」seal — no official revision history yet. */
export function isPersistedCompletedSeal(
  status: string | undefined | null,
  currentRevisionVersion: number | undefined | null,
): boolean {
  return (
    normalizeExaminationLockStatus(status) === "完了" &&
    currentRevisionVersion == null
  );
}

export function isPersistedResultsLocked(
  status: string | undefined | null,
  currentRevisionVersion: number | undefined | null,
): boolean {
  return (
    isPersistedConfirmedStatus(status) ||
    isPersistedCompletedSeal(status, currentRevisionVersion)
  );
}

File: backend/internal/medicalrecord/examination_lock.go
package medicalrecord

import (
	"github.com/animal-ekarte/backend/internal/apperrors"
	"github.com/animal-ekarte/backend/internal/model"
)

// examinationFullyLocked is the official seal: no parent or result mutation.
func examinationFullyLocked(exam *model.Examination) bool {
	return exam != nil && exam.Status == model.ExaminationStatusConfirmed
}

// examinationResultsLocked rejects result-row mutation and delete of sealed exams.
//
// - confirmed: always locked
// - completed without revision history: first-pass completion seal (BUG-033)
// - completed with revision history: post-unconfirm working copy — still editable
func examinationResultsLocked(exam *model.Examination) bool {
	if exam == nil {
		return false
	}
	switch exam.Status {
	case model.ExaminationStatusConfirmed:
		return true
	case model.ExaminationStatusCompleted:
		return exam.CurrentRevisionVersion == nil
	default:
		return false
	}
}

func errExaminationFullyLocked() error {
	return apperrors.WrapConflict("確定済みの検査は編集できません")
}

func errExaminationResultsLocked(exam *model.Examination) error {
	if exam != nil && exam.Status == model.ExaminationStatusConfirmed {
		return errExaminationFullyLocked()
	}
	return apperrors.WrapConflict("完了済みの検査結果は編集できません")
}

func errExaminationDeleteLocked(exam *model.Examination) error {
	if exam != nil && exam.Status == model.ExaminationStatusConfirmed {
		return apperrors.WrapConflict("確定済みの検査は削除できません")
	}
	return apperrors.WrapConflict("完了済みの検査は削除できません")
}
